src/main/java/com/mraof/minestuck/world/lands/gen/LandTypeSelectionLoader.java

```diff
@@ -0,0 +1,171 @@
+	
+	public static final String TERRAIN_PATH = "minestuck/terrain_land_types.json";
+	public static final String TITLE_PATH = "minestuck/title_land_types.json";
+	
+	public record RawData(List<GroupData> terrainTypeData, Map<EnumAspect, List<GroupData>> titleTypeData)
+	{}
+	
+	private record GroupData(Either<List<ResourceLocation>, ExtraCodecs.TagOrElementLocation> value)
+	{
+		private <A> Optional<LandTypeSelection.Group<A>> lookup(IForgeRegistry<A> registry)
+		{
+			return this.value.map(list -> {
+				List<A> elements = new ArrayList<>();
+				for(ResourceLocation id : list)
+					getOrLog(registry, id).ifPresent(elements::add);
+				
+				if(elements.isEmpty())
+					return Optional.empty();
+				else
+					return Optional.of(LandTypeSelection.Group.of(elements));
+			}, location -> {
+				if(location.tag())
+				{
+					TagKey<A> tag = TagKey.create(registry.getRegistryKey(), location.id());
+					return Optional.of(LandTypeSelection.Group.of(tag));
+				}
+				else
+					return getOrLog(registry, location.id()).map(landType -> LandTypeSelection.Group.of(Collections.singletonList(landType)));
+			});
+		}
+		
+		private static <A> Optional<A> getOrLog(IForgeRegistry<A> registry, ResourceLocation id)
+		{
+			A element = registry.getValue(id);
+			if(element == null)
+				LOGGER.error("Could not find land type for id in registry {}: {}", registry.getRegistryKey(), id);
+			return Optional.ofNullable(element);
+		}
+	}
+	
+	private static final Codec<GroupData> GROUP_DATA_CODEC = RecordCodecBuilder.create(instance -> instance.group(new ExtraCodecs.EitherCodec<>(ResourceLocation.CODEC.listOf(), ExtraCodecs.TAG_OR_ELEMENT_ID).fieldOf("value").forGetter(GroupData::value)).apply(instance, GroupData::new));
+	private static final Codec<List<GroupData>> TERRAIN_DATA_CODEC = GROUP_DATA_CODEC.listOf();
+	private static final Codec<Map<EnumAspect, List<GroupData>>> TITLE_DATA_CODEC = Codec.unboundedMap(EnumAspect.CODEC, GROUP_DATA_CODEC.listOf());
+	
+	@SubscribeEvent
+	public static void onResourceReload(AddReloadListenerEvent event)
+	{
+		event.addListener(new LandTypeSelectionLoader());
+	}
+	
+	@Override
+	protected RawData prepare(ResourceManager resourceManager, ProfilerFiller profiler)
+	{
+		List<GroupData> terrainData = new ArrayList<>();
+		
+		for(String namespace : resourceManager.getNamespaces())
+		{
+			ResourceLocation location = new ResourceLocation(namespace, TERRAIN_PATH);
+			resourceManager.getResource(location).ifPresent(resource -> {
+				try(Reader reader = resource.openAsReader())
+				{
+					JsonElement json = JsonParser.parseReader(reader);
+					List<GroupData> parsedData = TERRAIN_DATA_CODEC.parse(JsonOps.INSTANCE, json).getOrThrow(false, LOGGER::error);
+					terrainData.addAll(parsedData);
+				} catch(IOException ignored)
+				{
+				} catch(RuntimeException runtimeexception)
+				{
+					LOGGER.warn("Invalid json in data pack: '{}'", location.toString(), runtimeexception);
+				}
+			});
+		}
```

Comment:
Could you put a comment at the beginning with a brief description of what happens in the for loop

Reply:
Instead split it into functions that should have the same effect.